### PR TITLE
Add Pointed Dripstone drip support to FluidTypes

### PR DESCRIPTION
**Pull request by GizmoTheMoonPig**
In 1.17, Mojang added Pointed Dripstone, which has a unique quirk in that it can fill cauldrons below it with fluids if the fluid is sitting above it. However, this system is extremely hardcoded. To get around it, you'll need to add 4 mixins to your project. (3 in `PointedDripstoneBlock` and 1 in `CauldronBlock`). This PR aims to allow modders to easily allow their fluids to drip from pointed dripstones if they so desire. 

This PR adds a new `FluidType` property, `addDripstoneDripping`. You can define the fill chance, the particle that drips from the tip of the pointed dripstone when your fluid is above it, what filled cauldron you want to replace an empty one with when the pointed dripstone successfully drips, and a sound to play when an empty cauldron fills up. Since this has been moved to FluidType properties, the hardcoded vanilla properties have been removed from `PointedDripstoneBlock` and `CauldronBlock` and moved to the fluid type registration for the fluids in `ForgeMod`.

This PR expands on an existing test mod and adds dripstone dripping to it. Placing `TEST_FLUID` from `FluidTypeTest` above a dripstone will spawn soul particles, fill the cauldron with a layer of powdered snow, and play the end portal opening sound when the cauldron is filled.

Please let me know if you guys aren't happy with how anything is named or how the Javadocs are worded. I'm not the best at wording things so I'm happy to change anything as needed.

### CODE REVIEW

File: patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch]: This does not need such a big change. It can just be an if statement at the top of the file with a redirect to a Forge hook. No need to replace the water and lava logic directly.

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch]: This doesn't need to be remove since it'll just go unused.

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:24]: This chance condition is backwards if you're describing it as the chance the cauldron will be filled. That should be `<`.

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:41]: Should we force a drip particle here?

Additionally, I don't know why there needs to be a separate particle by default when we can just have a single particle which can take in a fluid with the particle options to grab the correct tint. This is my opinion though, and I don't think such overengineering is completely necessary.

File: src/main/java/net/minecraftforge/fluids/FluidType.java
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/fluids/FluidType.java]: What does null mean then? Also, nullable can be assumed from the method signature.

File: src/main/java/net/minecraftforge/fluids/FluidType.java
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/fluids/FluidType.java:1180]: You should be using a sound action for this. Probably something like `CAULDRON_DRIP`. It can be abstracted away in the body of the method.

File: src/main/java/net/minecraftforge/fluids/FluidType.java
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/fluids/FluidType.java]: Javadoc should be here as well.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
So, this eventually brings in a separate question of should we provide defaults for cauldrons and particles that are dynamic for the fluid stored, but for right now, I'm fine with the proposed solution.

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (GizmoTheMoonPig) [reply patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch]: I'll have to at least remove the part that makes it return if the fluid isn't lava since that will make the patch effectively useless, but I'll add back the other stuff

File: patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch
Comment (GizmoTheMoonPig) [reply patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch]: I figured the patch would look cleaner in the code if I did it this way, but I'll change it if you guys prefer it that way :)

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (GizmoTheMoonPig) [reply patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:24]: I just replaced the old chance float with the one from the dripstone info here. The code works the same way it should (in that it checks that the random float is less than the chance, it's just that the original code uses greater than and inverts it. I can change this if desired, just wanted to keep it consistent.

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (GizmoTheMoonPig) [reply patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:41]: I'll make the particle field nullable. 
Personally, I don't think forcing people to use a drip particle is the way to go about this. As it currently stands you can spawn any particle you want to. While it doesn't make the most sense to do so in this context, it allows modders to freely customize how their fluid looks when it drips from a dripstone. Thats my stance though. This isn't my project though so i'll do what you guys prefer

File: src/main/java/net/minecraftforge/fluids/FluidType.java
Comment (GizmoTheMoonPig) [reply src/main/java/net/minecraftforge/fluids/FluidType.java:1180]: Will switch to a SoundAction. 
I initially wanted to keep the SoundEvent in the record to make it less confusing to add one, but I suppose it makes more sense if it's a SoundAction since that's how all other fluid sounds are handled

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (ChampionAsh5357) [reply patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:24]: That's fine, but the javadocs should be updated then to reflect that.

File: patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch
Comment (ChampionAsh5357) [reply patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch]: Let me confirm on this quickly to see if we're still prioritizing minimal changes

Comment (ChampionAsh5357) [reply patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch]: Ok, we still are prioritizing the least intrusive approach.

File: patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/CauldronBlock.java.patch]: This should probably be in a `ForgeHooks` method. Also the else can be moved to a separate line to avoid the deletion.

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch]: Wouldn't this just default to water? It may be better to just do an if not null check on the line below.

File: src/main/java/net/minecraftforge/fluids/FluidType.java
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/fluids/FluidType.java:1177]: This should be updated to reflect the inverted chance check.

File: src/main/java/net/minecraftforge/fluids/FluidType.java
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/fluids/FluidType.java]: You can still have an overload that takes in the fill sound and then just redirect that to the sound action.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (ChampionAsh5357) [reply patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:24]: Ignore me, I missed the negation sign for some reason.

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:41]: Apologies if I wasn't being clear. I was saying that if drip info is null, then the particle should be null as well and not default to the water particle

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch
Comment (GizmoTheMoonPig) [reply patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:41]: Vanilla spawns water particles on the dripstone tip even if there's no fluid above it. Should this behavior be mimicked if the fluid above doesn't have and drip info, or should it just do nothing? Personally, I think that it should still spawn water particles if there's no drippable fluid above, essentially mimicking what vanilla does when anything but a fluid is above it. Totally up to you though, just wanted to point that out.

Comment (ChampionAsh5357) [reply patches/minecraft/net/minecraft/world/level/block/PointedDripstoneBlock.java.patch:41]: I just did a quick check on the bug tracker, and mojang says it works as intended. So, no changes needed, apologies.

**Review by ChampionAsh5357 — APPROVED** [review]
@neoforged/maintainers please test PR before merging. Codewise, it looks fine, but I can't verify functionality.

Check custom fluid behavior along with water and lava. Also verify that without the parameters set the default logic is still executed properly.

**Comment by GizmoTheMoonPig** [comment]
Was having a bit of trouble updating this PR for 1.20.2, so im just gonna open a new one with the changes

**Pull request closed without merging**
